src/bson.rs

```diff
@@ -366,12 +415,54 @@ impl From<Bson> for Value {
             Bson::DbPointer(DbPointer {
                 ref namespace,
                 ref id,
-            }) => json!({ "$dbPointer": { "$ref": namespace, "$id": id.to_string() } }),
+            }) => json!({
+                "$dbPointer": {
+                    "$ref": namespace,
+                    "$id": {
+                        "$oid": id.to_hex()
+                    }
+                }
+            }),
+        }
+    }
+
+    pub fn into_canonical_extjson(self) -> Value {
+        match self {
+            Bson::I32(i) => json!({ "$numberInt": i.to_string() }),
```

Comment:
Most cases are identical in canonical JSON to relaxed; the exceptions are numbers, datetimes, and from a recursive perspective, arrays and documents, since they could contain values that are numbers or datetimes.